### PR TITLE
[CMSIS-NN] Convert scalar constants to tensor constants

### PR DESCRIPTION
**Pull request by ashutosh-arm**
* Scalar to tensor constant pass

-Pass that converts scalar constants into tensor constants for CMSIS-NN binary ops
-Updated extract_constants pass to support constants at any position in the function signature
-Modified tests for extract_constants pass corresponding to the scalar support
-Added new binary op tests to include constants

### CODE REVIEW

File: src/relay/backend/contrib/cmsisnn/extract_constants.cc [hunk 1]
```diff
@@ -62,17 +62,87 @@ class ExtractConstantsMutator : public MixedModeMutator {
...
 
+  // Creates new arguments from current call's arguments
+  // Updates constants into the caller arguments: here caller signifies caller that comprises call
+  // to func
+  Array<Expr> CreateNewCallArgsFromKickedoutConstants(Call call, Function func) {
+    ICHECK(function_to_arguments_.find(func) != function_to_arguments_.end());
+    Array<Expr> fSignature(function_to_arguments_[func]);
+
+    // Is func a global_function?
+    // main() is not registered for kicking out constants
```
Comment (Mousius) [inline src/relay/backend/contrib/cmsisnn/extract_constants.cc:84]:
```suggestion
    // main() is not registered for extracting constants
```

File: src/relay/backend/contrib/cmsisnn/extract_constants.cc [hunk 2]
```diff
@@ -62,17 +62,87 @@ class ExtractConstantsMutator : public MixedModeMutator {
...
     }
     return std::move(func);
   }
 
+  // Creates new arguments from current call's arguments
+  // Updates constants into the caller arguments: here caller signifies caller that comprises call
+  // to func
+  Array<Expr> CreateNewCallArgsFromKickedoutConstants(Call call, Function func) {
+    ICHECK(function_to_arguments_.find(func) != function_to_arguments_.end());
+    Array<Expr> fSignature(function_to_arguments_[func]);
```
Comment (Mousius) [inline src/relay/backend/contrib/cmsisnn/extract_constants.cc:81]:
```suggestion
    Array<Expr> function_signature(function_to_arguments_[func]);
```

File: src/relay/backend/contrib/cmsisnn/extract_constants.cc [hunk 3]
```diff
@@ -62,17 +62,87 @@ class ExtractConstantsMutator : public MixedModeMutator {
...
+
+    // Is func a global_function?
+    // main() is not registered for kicking out constants
+    bool is_global_function = functions_.empty() ? true : false;
+
+    bool new_constants_added = false;
+    // This tracks arguments traversed inside fSignature
+    uint32_t fsignature_id = 0;
```
Comment (Mousius) [inline src/relay/backend/contrib/cmsisnn/extract_constants.cc:89]:
```suggestion
    uint32_t function_signature_id = 0;
```

File: src/relay/backend/contrib/cmsisnn/extract_constants.cc [hunk 4]
```diff
@@ -62,17 +62,87 @@ class ExtractConstantsMutator : public MixedModeMutator {
...
+          << "Rest of the collected arguments should be constant in the partitioned function.";
+      new_caller_args.push_back(GetRef<Constant>(constant));
+      new_call_args.push_back(Var(gen_var_name(), constant->tensor_type()));
+      new_constants_added = true;
+    }
+
+    // Update the arguments of caller of local function
+    if (new_constants_added && !is_global_function) {
+      const Function& last_func = functions_.back();
+      Array<Expr> fconstants(function_to_arguments_[last_func]);
```
Comment (Mousius) [inline src/relay/backend/contrib/cmsisnn/extract_constants.cc:137]:
```suggestion
      Array<Expr> function_constants(function_to_arguments_[last_func]);
```

File: src/relay/backend/contrib/cmsisnn/scalar_to_tensor_constant.cc [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+namespace tvm {
+namespace relay {
+namespace contrib {
+namespace cmsisnn {
+
+/*!
+ * \brief This Mutator finds all partitioned functions meant for CMSIS-NN binary ops.
+ * Then, it substitutes the scalar constants with tensor constants. It makes the shape of this
+ * new constant same as that of the neighbouring constant of the other binary operand. The
+ * expectation is that the ExtractConstant pass would later kick this tensor constant out of the
```
Comment (Mousius) [inline src/relay/backend/contrib/cmsisnn/scalar_to_tensor_constant.cc:43]:
```suggestion
 * expectation is that the ExtractConstant pass would later extract the tensor constant out of the
```

File: src/relay/backend/contrib/cmsisnn/scalar_to_tensor_constant.cc [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+  //     qnn.mul(%input0, %input1, scale0, zero_point0,
+  //              scale1, zero_point_1, output_scale, output_zero_point)
+  //   };
+  //   %56(meta[relay.Constant] /* _scalar constant_ */, %cmsis-nn_input)
+  // }
+  Expr Rewrite_(const CallNode* call, const Expr& post) final {
+    Expr final_call = post;
+    call = post.as<CallNode>();
+
+    // Create a new variable argument that is of the same shape as the neibouring argument
```
Comment (Mousius) [inline src/relay/backend/contrib/cmsisnn/scalar_to_tensor_constant.cc:70]:
```suggestion
    // Create a new variable argument that is of the same shape as the neighbouring argument
```

File: src/relay/backend/contrib/cmsisnn/scalar_to_tensor_constant.cc [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+        String arg_name = arg.as<VarNode>()->name_hint();
+        int tensor_arg_id = (i + 1) % 2;
+        Expr tensor_arg = call->args[tensor_arg_id];
+        if (!tensor_arg->checked_type_.defined()) {
+          continue;
+        }
+        TensorType tensor_type = GetRef<TensorType>(tensor_arg->type_as<TensorTypeNode>());
+        new_args.Set(i, Var(arg_name, tensor_type));
+      }
+      final_call = Call(call->op, new_args, call->attrs, {});
```
Comment (Mousius) [inline src/relay/backend/contrib/cmsisnn/scalar_to_tensor_constant.cc:94]: Can these larger blocks inside of the if-statements be extracted into functions for clarity?

File: src/relay/backend/contrib/cmsisnn/scalar_to_tensor_constant.cc [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+  IRModule mod_;
+};
+
+IRModule ScalarToTensorConstant(const IRModule& mod) {
+  auto mutator = ScalarToTensorConstantMutator(mod);
+  Function main_func = Downcast<Function>(mod->Lookup("main"));
+  auto new_main_body = mutator.VisitExpr(main_func->body);
+  if (!new_main_body.same_as(main_func->body)) {
+    auto main_var = mod->GetGlobalVar("main");
+    auto new_main_func = Function(main_func->params, new_main_body, main_func->ret_type,
```
Comment (Mousius) [inline src/relay/backend/contrib/cmsisnn/scalar_to_tensor_constant.cc:169]: Should this use `WithFields` ?

File: tests/python/contrib/test_cmsisnn/test_binary_ops.py [hunk 9]
```diff
@@ -46,12 +54,26 @@ def make_model(
     input_1_zero_point,
     out_scale=1.0 / 256,
     out_zero_point=-128,
+    input_0_type=BinaryOpInputType.Variable,
+    input_1_type=BinaryOpInputType.Variable,
 ):
     """Create a Relay Function / network model"""
 
+    def create_input(name, input_type, shape, dtype):
```
Comment (Mousius) [inline tests/python/contrib/test_cmsisnn/test_binary_ops.py:62]: Does this have to be nested inside `make_model`? We should also consider if we want this many branches in our test functions, you could parameterise with input creation functions to flatten out the logic.

File: tests/python/contrib/test_cmsisnn/test_binary_ops.py [hunk 10]
```diff
@@ -131,6 +153,140 @@ def test_op_int8(op, input_0_scale, input_0_zero_point, input_1_scale, input_1_z
...
+    assert any(attrs), "At least one function with external attributes was expected."
+
+    compilers = [
+        key == "Compiler" and value == "cmsis-nn" for attr in attrs for key, value in attr.items()
+    ]
+    assert any(compilers), "Module does not contain function for cmsisnn target."
+
+    assert count_num_calls(orig_mod) == count_num_calls(
+        cmsisnn_mod
+    ), "Number of calls changed during partitioning"
```
Comment (Mousius) [inline tests/python/contrib/test_cmsisnn/test_binary_ops.py:234]: We're re-using this across tests, should we store it in an `assert_partitioned` or similar?

File: tests/python/contrib/test_cmsisnn/test_binary_ops.py [hunk 11]
```diff
@@ -131,6 +153,140 @@ def test_op_int8(op, input_0_scale, input_0_zero_point, input_1_scale, input_1_z
...
+
+    orig_mod = make_module(model)
+    cmsisnn_mod = cmsisnn.partition_for_cmsisnn(orig_mod)
+
+    attrs = [
+        cmsisnn_mod[var.name_hint].attrs
+        for var in cmsisnn_mod.get_global_vars()
+        if cmsisnn_mod[var.name_hint].attrs
+    ]
+    assert not any(attrs), "No function should have an external attribute."
```
Comment (Mousius) [inline tests/python/contrib/test_cmsisnn/test_binary_ops.py:287]: Similar to the above, can we put this in a function for re-use?

**Review by Mousius — CHANGES_REQUESTED** [review]
First pass over this @ashutosh-arm, thanks for fixing this :smile_cat:

File: src/relay/backend/contrib/cmsisnn/extract_constants.cc (hunk 2 above)
Comment (ashutosh-arm) [reply src/relay/backend/contrib/cmsisnn/extract_constants.cc:81]: ACK

File: src/relay/backend/contrib/cmsisnn/scalar_to_tensor_constant.cc (hunk 7 above)
Comment (ashutosh-arm) [reply src/relay/backend/contrib/cmsisnn/scalar_to_tensor_constant.cc:94]: Sure, that would make sense

File: src/relay/backend/contrib/cmsisnn/scalar_to_tensor_constant.cc (hunk 8 above)
Comment (ashutosh-arm) [reply src/relay/backend/contrib/cmsisnn/scalar_to_tensor_constant.cc:169]: yes, will do.

File: tests/python/contrib/test_cmsisnn/test_binary_ops.py (hunk 9 above)
Comment (ashutosh-arm) [reply tests/python/contrib/test_cmsisnn/test_binary_ops.py:62]: As discussed offline, I would make the change to parameterize it.

File: tests/python/contrib/test_cmsisnn/test_binary_ops.py (hunk 10 above)
Comment (ashutosh-arm) [reply tests/python/contrib/test_cmsisnn/test_binary_ops.py:234]: Can we do this as a separate PR as this applies to all CMSIS-NN tests?

File: tests/python/contrib/test_cmsisnn/test_binary_ops.py (hunk 11 above)
Comment (ashutosh-arm) [reply tests/python/contrib/test_cmsisnn/test_binary_ops.py:287]: Separate PR?

**Comment by ashutosh-arm** [comment]
Thanks for the review @Mousius. I have updated the PR with review comments.

File: tests/python/contrib/test_cmsisnn/test_binary_ops.py (hunk 10 above)
Comment (Mousius) [reply tests/python/contrib/test_cmsisnn/test_binary_ops.py:221]: Sounds good to me!

File: tests/python/contrib/test_cmsisnn/test_binary_ops.py (hunk 11 above)
Comment (Mousius) [reply tests/python/contrib/test_cmsisnn/test_binary_ops.py:271]: Sounds good to me!

**Review by Mousius — APPROVED** [review]
LGTM!

**Comment by Mousius** [comment]
Thanks @ashutosh-arm! This is merged now :smile_cat:

**Pull request merged**